tracer/src/Datadog.Trace/Debugger/Instrumentation/Registry/ProbeMetadataRegistry.cs

```diff
@@ -0,0 +1,93 @@
+// <copyright file="ProbeMetadataRegistry.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.Debugger.Expressions;
+using Datadog.Trace.Debugger.Helpers;
+using Datadog.Trace.Debugger.RateLimiting;
+
+namespace Datadog.Trace.Debugger.Instrumentation.Registry
+{
+    /// Acts as a registry of indexed <see cref="ProbeMetadataInfo"/>.
+    /// Each instrumented probe is given an index (hard-coded into the instrumented bytecode),
+    /// and this index is used to perform O(1) lookup for the corresponding ProbeMetadataInfo instance.
+    /// The reason the index is incremented and dictated by the native side is to support multi-AppDomain scenarios.
+    /// In these scenario, there will be multiple <see cref="EverGrowingRegistry{TPayload}.Items"/> arrays, one for each AppDomain.
+    /// In order for us to grab the same ProbeMetadataInfo across all of them, we need to dereference the same index,
+    /// because the same instrumented bytecode could execute in different AppDomains, and static fields are not shared across AppDomains.
+    internal class ProbeMetadataRegistry : EverGrowingRegistry<ProbeMetadataInfo>
+    {
+        private static ProbeMetadataRegistry _instance;
+        private static object _instanceLock = new();
+        private static bool _instanceInitialized;
+
+        internal static ProbeMetadataRegistry Instance
+        {
+            get
+            {
+                return LazyInitializer.EnsureInitialized(
+                    ref _instance,
+                    ref _instanceInitialized,
+                    ref _instanceLock);
+            }
+        }
+
+        /// <summary>
+        /// Tries to create a new <see cref="ProbeMetadataInfo"/> at <paramref name="index"/>.
+        /// </summary>
+        /// <param name="index">The index of the probe inside <see cref="EverGrowingRegistry{TPayload}.Items"/></param>
+        /// <param name="probeId">The id of the probe/></param>
+        /// <returns>true if succeeded (either existed before or just created), false if fails to create</returns>
+        public bool TryCreateProbeMetadataIfNotExists(int index, string probeId)
```

Comment:
nit: There is a bit of code duplication between this and the other `TryCreate*IfNotExists`, maybe we could extract the code that creates the thing to an abstract method to improve it slightly?

Reply:
https://github.com/DataDog/dd-trace-dotnet/pull/3725#discussion_r1097559895

Reply:
Fair enough - resolving.